File: handy-python/quiz/q16_best_time_to_buy_and_sell_stock_easy.py
"""
Q16) Best Time to Buy and Sell Stock

AI-BEST:
- Track the minimum price so far and the best profit.
- Time: O(n), Space: O(1)
"""


def max_profit_best(prices: list[int]) -> int:
    if not prices:
        return 0

    # Step 1: Track the cheapest price seen so far.
    # Why: The best sell decision on any day depends on the cheapest earlier buy.
    min_price = prices[0]
    best_profit = 0

    # Step 2: Compute the profit for selling on each day.
    # Why: Every day is a candidate sell day once we know the best prior buy.
    for price in prices[1:]:
        best_profit = max(best_profit, price - min_price)
        min_price = min(min_price, price)

    # Step 3: Return the best profit found.
    # Why: We have already checked every buy/sell pairing implicitly.
    return best_profit


def run_demo() -> None:
    print("Q16: Best Time to Buy and Sell Stock")
    samples = [
        [7, 1, 5, 3, 6, 4],
        [7, 6, 4, 3, 1],
        [1, 2],
        [],
    ]
    for prices in samples:
        print(f"Input: prices={prices}")
        print("  BEST:", max_profit_best(prices))


if __name__ == "__main__":
    run_demo()

File: handy-ts/quiz/q16_best_time_to_buy_and_sell_stock_easy.ts
/**
 * Q16) Best Time to Buy and Sell Stock
 *
 * AI-BEST:
 * Track the minimum price so far and the best profit so far.
 * Time: O(n), Space: O(1)
 */

function maxProfitBest(prices: number[]): number {
  // Step 1: Track the cheapest price seen so far because any good sale depends on a prior buy.
  let minPrice = Number.POSITIVE_INFINITY;
  let bestProfit = 0;

  // Step 2: Evaluate profit at each day because selling today only depends on the cheapest earlier day.
  for (const price of prices) {
    minPrice = Math.min(minPrice, price);
    bestProfit = Math.max(bestProfit, price - minPrice);
  }

  // Step 3: Return the largest spread because that is the best single transaction allowed.
  return bestProfit;
}

function runDemo(): void {
  console.log("Q16: Best Time to Buy and Sell Stock");
  const cases = [
    [7, 1, 5, 3, 6, 4],
    [7, 6, 4, 3, 1],
    [1, 2, 3, 4, 5],
  ];

  for (const prices of cases) {
    console.log(`Input: prices=${JSON.stringify(prices)}`);
    console.log("  BEST:", maxProfitBest(prices));
  }
}

runDemo();

export {};
